pkg/kv/kvserver/store_split.go

```diff
@@ -105,9 +85,24 @@ func splitPreApply(
 				log.Fatalf(ctx, "failed to load hard state for removed rhs: %v", err)
 			}
 		}
-		const rangeIDLocalOnly = false
-		const mustUseClearRange = false
-		if err := clearRangeData(&split.RightDesc, readWriter, readWriter, rangeIDLocalOnly, mustUseClearRange); err != nil {
+		if err := clearRangeData(split.RightDesc.RangeID, readWriter, readWriter, clearRangeDataOptions{
+			// We know there isn't anything in these two replicated spans below in the
+			// right-hand side (before the current batch), so setting these options
+			// will in effect only mask the writes to the RHS replicated state we have
+			// staged in this batch, which is what we're after.
+			ClearReplicatedBySpan:    split.RightDesc.RSpan(),
+			ClearReplicatedByRangeID: true,
+			// See the HardState write-back dance above and below.
+			//
+			// TODO(tbg): we don't actually want to touch the raft state of the right
+			// hand side replica since it's absent or a more recent replica than the
+			// split. Now that we have a boolean targeting the unreplicated
+			// RangeID-based keyspace, we can set this to false and remove the
```

Comment:
We shouldn't touch the RHS raft state here at all. The only reason we do this writeback is because the previous API forced us to remove everything. The RHS is for all intents and purposes a completely unrelated replica in this context, so there shouldn't be a reason to touch it.
But you have a good point that it's not as simple as that. We are operating here on the `WriteBatch` that the split is using. We really are trying to remove writes from that WriteBatch, rather than trying to remove anything from the engine. But there isn't a way to "delete just from the batch but not from the engine the batch sits on top of". So the question is, what is in the `WriteBatch`?

The relevant comment is here:

https://github.com/cockroachdb/cockroach/blob/f263a765d750e41f2701da0a923a6e92d09159fa/pkg/kv/kvserver/batcheval/cmd_end_transaction.go#L1109-L1149

The long and short of it is that we don't write to the RangeID-unreplicated key space in the WriteBatch - we rely on the call to `SynthesizeRaftState` (which only happens further down below, and which we won't reach in this case since we're returning early).

Expanded the comment a bit with this information.